File: programmers/level2/조이스틱.java
package programmers;

public class 조이스틱 {
	static
	class Solution {
	    public int solution(String name) {
	        int answer = 0;
	        StringBuilder sb = new StringBuilder();
	        for(int i = 0; i < name.length(); ++i) {
	        	sb.append("A");
	        }
	        answer = dfs(sb,name,0,0);
	        return answer;
	    }
	    int dfs(StringBuilder sb, String name, int count, int currentIdx) {
	    	if(sb.toString().equals(name)) {
	    		return count;
	    	}
	    	int rCount = Integer.MAX_VALUE;
	    	int lCount = Integer.MAX_VALUE;
	    	int rIdx = findRightIdx(currentIdx, sb, name);
	    	int lIdx = findLeftIdx(currentIdx, sb, name);
	    	if(rIdx != -1) {
	    		char currentChar = sb.charAt(rIdx);
	    		int d = diff(sb.charAt(rIdx), name.charAt(rIdx), 26);
	    		sb.setCharAt(rIdx, name.charAt(rIdx));
	    		rCount = dfs(sb,name,count+diff(currentIdx, rIdx, name.length())+d,rIdx);
	    		sb.setCharAt(rIdx, currentChar);
	    	}
	    	if(lIdx != -1) {
	    		char currentChar = sb.charAt(lIdx);
	    		int d = diff(sb.charAt(lIdx), name.charAt(lIdx), 26);
	    		sb.setCharAt(lIdx, name.charAt(lIdx));
	    		lCount = dfs(sb,name,count+diff(currentIdx, lIdx,  name.length())+d,lIdx);
	    		sb.setCharAt(lIdx, currentChar);
	    	}
	    	return Math.min(rCount, lCount);
	    }
	    int diff(int a, int b, int size) {
	    	if(a > b) {
	    		int temp = a;
	    		a = b;
	    		b = temp;
	    	}
	    	return Math.min(b-a, a+size-b);
	    }
	    int findRightIdx(int currentIdx, StringBuilder sb, String name) {
	    	for(int i = 0; i < name.length(); ++i) {
	    		int idx = currentIdx + i;
	    		if(idx >= name.length()) {
	    			idx -= name.length();
	    		}
	    		if(sb.charAt(idx) != name.charAt(idx)) {
	    			return idx;
	    		}
	    	}
	    	return -1;
	    }
	    int findLeftIdx(int currentIdx, StringBuilder sb, String name) {
	    	for(int i = 0; i < name.length(); ++i) {
	    		int idx = currentIdx - i;
	    		if(idx < 0) {
	    			idx += name.length();
	    		}
	    		if(sb.charAt(idx) != name.charAt(idx)) {
	    			return idx;
	    		}
	    	}
	    	return -1;
	    }
	}
	public static void main(String[] args) {
		String name = "JEROEN";
		Solution sol = new Solution();
		System.out.println(sol.solution(name));
	}
}

File: programmers/python/level2/조이스틱.py
answer = 987654321
def solution(name):
    dfs(0,'A'*len(name), name,0)
    return answer

def dfs(index, string, name, count):
    diff = get_diff('A', name[index])
    count += diff
    string = string[:index]+name[index]+string[index+1:]
    if string == name:
        global answer
        if count < answer:
            answer = count
        return
    
    next = left_move(index-1, string, name)
    dfs(next[0], string , name, count+next[1])

    next = right_move(index+1, string, name)
    dfs(next[0], string , name, count+next[1])

def left_move(index, string, name):
    count = 1
    if index < 0:
        index += len(string)
    while string[index] == name[index] and count < len(string):
        index -= 1
        count += 1
        if index < 0:
            index += len(string)
    return (index, count)

def right_move(index, string, name):
    count = 1
    if index >= len(string):
        index = 0
    while string[index] == name[index] and count < len(string):
        index += 1
        count += 1
        if index >= len(string):
            index = 0
    return (index, count)

def get_diff(a, b):
    if a > b:
        a,b = b,a
    diff1 = ord(b)-ord(a)
    diff2 = ord(a)+26 - ord(b)
    return min(diff1, diff2)

print(solution('JAN'))